README.md
# MMM-public-transport-gtfs

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Aggregates compressed [GTFS static](https://developers.google.com/transit/gtfs) (General Transit Feed Specification) data and displays upcoming departure times of chosen public transportation stops.<br>

Currently, only tested with gtfs.zip provided by ["Rīgas satiksme"](https://www.rigassatiksme.lv/en/) and some randomly chosen GTFS sources.<br>
_Maybe worth adding list of tested sources, hmm_

![](https://raw.githubusercontent.com/DanielsSt/MMM-public-transport-gtfs/master/screenshots/Screenshot.png)

## Using the module

To use this module:<br>
run these commands from MagicMirror directory (usually `~/MagicMirror`), to clone the repo and install dependencies
```shell
cd modules
git clone https://github.com/DanielsSt/MMM-public-transport-gtfs
cd MMM-public-transport-gtfs
npm install
```

afterwards add the following configuration block to the modules array in the `config/config.js` file and (re)start MagicMirror:
```js
var config = {
    modules: [
        {
            module: 'MMM-public-transport-gtfs',
            config: {
                // See below for configurable options
            }
        }
    ]
}
```

## Configuration options

| Option           | Description                                               
|----------------- |-----------------------------------------------------------
| `gtfsDataUrl`        | *Required* URL to zipped GTFS data <br><br> **Type:** `string`(URL) <br> *Example* `https://saraksti.rigassatiksme.lv/riga/gtfs.zip`                         
| `stops`  | *Required* stop_ids of stops you want to display <br>(can be found on `stops.txt`)<br><br>**Type:** `string[]` <br>Default `[]`
| `domUpdateInterval`  | *Optional* Time between content updates on screen <br><br>**Type:** `int`(milliseconds) <br>Default `60000` milliseconds (1 minute)
| `dataUpdateInterval`  | *Optional* Time between GTFS data updates<br><br>**Type:** `int`(milliseconds) <br>Default `86400000` milliseconds (1 day)
| `display`  | *Optional* How many upcoming transports to display<br><br>**Type:** `int` <br>Default `10`
| `use24HrClock`  | *Optional* Display 24hr clock<br><br>**Type:** `bool` <br>Default `true`

### Example config

```js
config: {
    // url to GTFS data
    gtfsDataUrl: "https://saraksti.rigassatiksme.lv/riga/gtfs.zip",
    // update screen every 30 seconds
    domUpdateInterval: 30000,
    // update GTFS data daily
    dataUpdateInterval: 86400000,
    // show next 5 transports
    display: 5,
    // display time in 24hr format, or set to false - for 12hr clock
    use24HrClock: true,
    // find your stops in stops.txt
    stops: ["0709", "0079", "0075", "7980", "2017"]
}
```

## Useful resources

* [GTFS Static Overview](https://developers.google.com/transit/gtfs)
* [OpenMobilityData](https://transitfeeds.com/)
* [MagicMirror²](https://github.com/MichMich/MagicMirror/)

## Known issues
* Only one GTFS data source can be used, maybe I will implement a way to provide and aggregate multiple GTFS sources;
* Startup takes ~15 seconds, most of the time is spent to just download and unzip the GTFS archive;
* Results are not cached, meaning, data is downloaded and parsed after every restart;
* Noticed that sometimes departure times are 1 minute off of officially available information and some shady practices of GTFS (like custom route types) are used;
* Icons displayed next to the route are not complete set, since font-awesome does not provide icon for every possible transport;


CHANGELOG.md
# MMM-public-transport-gtfs Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Unreleased

First public release


node_helper.js
var NodeHelper = require("node_helper");
var JSZip = require("jszip");
var axios = require("axios");
var Log = require("logger");
const parse = require("csv-parse/lib/sync");
const {keys} = require("grunt/lib/grunt/option");

module.exports = NodeHelper.create({
	stops: [],
	url: "",
	gtfsFiles: [
		"stops.txt",
		"stop_times.txt",
		"trips.txt",
		"routes.txt",
		"calendar.txt",
		"calendar_dates.txt",
	],
	skippableGtfsFiles: [
		"calendar_dates.txt",
	],

	socketNotificationReceived: function(notification, payload) {
		if (notification === this.name + "-GET_GTFS") {
			this.stops = payload.stops;
			this.url = payload.url;

			if (this.url === null || this.url === ""){
				self.logError("URL not provided", "ERROR_NO_GTFS_URL");
				return;
			}

			if (this.stops.length === 0){
				self.logError("No stops provided", "ERROR_NO_STOPS");
				return;
			}

			this.getGtfsData();
		}
	},

	getGtfsData: function() {
		var self = this;
		var urlApi = this.url;

		// reset unzipped files
		this.unzippedGtfsData = {};
		this.filesUnzipped = 0;

		Log.info(self.name, "Pulling GTFS data", urlApi);
		axios.get(urlApi, {responseType: "arraybuffer"})
			.then(function(response) {
				if (response.status === 200) {
					Log.info(self.name, "GTFS data downloaded, unzipping");
					self.unzipGtfsData(response.data);
				} else {
					self.logError("Could not load data.", "ERROR_FAILED_DL", response.status);
				}
			}).catch(function(error){
				self.logError("Could not load data.!", "ERROR_FAILED_DL",
					{ message: error.message, stack: error.stack }
				);
				throw error;
			});
	},

	unzippedGtfsData: {},

	unzipGtfsData: function(zipContents) {
		var self = this;
		JSZip.loadAsync(zipContents).then(function (zip) {
			var contents = {};
			zip.forEach(function (relativePath, zipEntry) {
				if (self.gtfsFiles.indexOf(zipEntry.name) >= 0){
					contents[zipEntry.name] = zip.file(zipEntry.name).async("string");
				}
			});

			var contentKeys = Object.keys(contents);
			if (contentKeys.length !== self.gtfsFiles.length) {
				var missingSkippableFiles = 0;
				self.skippableGtfsFiles.forEach(function(filename, index, array){
					if(contentKeys.indexOf(filename) === -1) {
						missingSkippableFiles++;
					}
				});

				if ((contentKeys.length + missingSkippableFiles) !== self.gtfsFiles.length) {
					self.logError("Did not find all required files in Zip", "ERROR_INVALID_FILES",
						{expected: self.gtfsFiles, got: Object.keys(contents)}
					);
					throw new Error("Did not find all required files in Zip");
				}
			}

			return contents;
		}).then(function (contents) {
			self.gtfsFiles.forEach(function (filename, index, array) {
				self.unzippedGtfsData[filename.replace(".txt", "")] = [];
				if (typeof contents[filename] === "undefined"){
					self.unzipGtfsSync();
					return;
				}

				contents[filename].then(function(fileContents){
					self.unzippedGtfsData[filename.replace(".txt", "")] = parse(fileContents.trim(), {
						columns: true,
						skip_empty_lines: true,
					});
				}).catch(function (error) {
					self.logError("Failed to parse ZIP contents!", "ERROR_PARSE_ZIP",
						{ message: error.message, stack: error.stack }
					);
					throw error;
				}).finally(function() {
					self.unzipGtfsSync();
				});
			});
		}).catch(function (error) {
			self.logError("Failed to process ZIP", "ERROR_PROCESS_ZIP",
				{ message: error.message, stack: error.stack }
			);
			throw error;
		});
	},

	// If anyone knows better way to synchronise this mess, be my guest
	filesUnzipped: 0,
	unzipGtfsSync: function() {
		this.filesUnzipped++;
		if (this.filesUnzipped >= Object.keys(this.unzippedGtfsData).length){
			Log.info(this.name, "GTFS data unzipped, processing data!");

			try {
				var processedData = this.processGtfsData(this.unzippedGtfsData);
			} catch (error){
				this.logError("Failed to process txt file contents!", "ERROR_PROCESS_TXT",
					{ message: error.message, stack: error.stack }
				);
				throw error;
			}

			this.sendSocketNotification(this.name + "-SET_GTFS", processedData);
		}
	},

	processGtfsData: function(data) {
		var self = this;
		var stops = {};

		data.stops.every(function(stop, index, array){
			if (self.stops.indexOf(stop.stop_id) >= 0) {
				stops[stop.stop_id] = stop;
			}
			return Object.keys(stops).length < self.stops.length;
		});

		var stopTimes = {};
		var tripIds = [];
		// find needed stops
		data.stop_times.forEach(function(stop, index){
			if (self.stops.indexOf(stop.stop_id) >= 0) {
				var stopTimesFields = ["arrival_time","departure_time"];
				stopTimesFields.forEach(function(field, index, array) {
					stop[field] = self.parseGtfsTime(stop[field]);
				});
				// group by hours
				var hour = stop.departure_time.getHours();
				if (typeof stopTimes[hour] === "undefined") {
					stopTimes[hour] = [];
				}
				stopTimes[hour].push(stop);

				if (tripIds.indexOf(stop.trip_id) === -1) {
					tripIds.push(stop.trip_id);
				}
			}
		});

		// sort times
		Object.keys(stopTimes).forEach(function (hour, index, array){
			stopTimes[hour] = stopTimes[hour].sort(function(a, b){
				return a.departure_time - b.departure_time;
			})
		});

		var trips = {};
		var serviceIds = [];
		var routeIds = [];
		// find trips going to the stops
		data.trips.every(function(trip, index){
			var tripIdIndex = tripIds.indexOf(trip.trip_id);
			if (tripIdIndex >= 0) {
				trips[trip.trip_id] = trip;
				tripIds.splice(tripIdIndex, 1);
				if (routeIds.indexOf(trip.route_id) === -1) {
					routeIds.push(trip.route_id);
				}

				if (serviceIds.indexOf(trip.service_id) === -1) {
					serviceIds.push(trip.service_id);
				}
			}
			return tripIds.length > 0;
		});

		var routes = {};
		// find the routes
		data.routes.every(function(route, index){
			var routeIdIndex = routeIds.indexOf(route.route_id);
			if (routeIdIndex >= 0) {
				routes[route.route_id] = route;
				routeIds.splice(routeIdIndex, 1);
			}
			return routeIds.length > 0;
		});

		var serviceCalendar = {};
		var serviceIds_ = Array.from(serviceIds);
		// find days when transport is(not) usually available
		data.calendar.every(function(calendarRecord, index){
			var serviceIdIndex = serviceIds_.indexOf(calendarRecord.service_id);
			if (serviceIdIndex >= 0) {
				serviceCalendar[calendarRecord.service_id] = calendarRecord;

				var dateColumns = ["start_date", "end_date"];
				dateColumns.forEach(function(dateColumn, index, array){
					serviceCalendar[calendarRecord.service_id][dateColumn] = self.parseGtfsDate(
						serviceCalendar[calendarRecord.service_id][dateColumn]);
				});

				var dayColumns = ["monday","tuesday","wednesday","thursday","friday","saturday","sunday"];
				dayColumns.forEach(function(day, index, array) {
					serviceCalendar[calendarRecord.service_id][day] = parseInt(serviceCalendar[calendarRecord.service_id][day]);
				});

				serviceIds_.splice(serviceIdIndex, 1);
			}
			return serviceIds_.length > 0;
		});

		var calendarExceptions = {};
		// find dates when there is an planned exception - either new service appears or is unavailable
		data.calendar_dates.forEach(function (date, index, array){
			if (serviceIds.indexOf(date.service_id) >= 0) {
				if (typeof calendarExceptions[date.service_id] === "undefined") {
					calendarExceptions[date.service_id] = [];
				}
				date.date = self.parseGtfsDate(date.date);
				calendarExceptions[date.service_id].push(date);
			}
		});

		return {
			stops: stops,
			stopTimes: stopTimes,
			trips: trips,
			routes: routes,
			serviceCalendar: serviceCalendar,
			calendarExceptions: calendarExceptions,
		};
	},

	parseGtfsDate: function (dateStr) {
		return new Date(
			parseInt(dateStr.substr(0, 4)),
			parseInt(dateStr.substr(4, 2)) - 1,
			parseInt(dateStr.substr(6, 2))
		);
	},

	parseGtfsTime: function (timeStr) {
		return new Date(
			1970,
			0,
			1,
			parseInt(timeStr.substr(0, 2)),
			parseInt(timeStr.substr(3, 2)),
			parseInt(timeStr.substr(6, 2))
		);
	},

	logError: function(message, code, additionalData= null) {
		Log.error(this.name, code, message, additionalData);
		this.sendSocketNotification(this.name + "-SET_NOTIFICATIONS",
			{
				code: code,
				additionalData: (additionalData !== null ? "<br><pre>" + JSON.stringify(additionalData, null, 2) + "<pre>" : "")
			}
		);
	},

});


MMM-public-transport-gtfs.js
Module.register("MMM-public-transport-gtfs", {
	defaults: {
		gtfsDataUrl: "",
		// every minute
		domUpdateInterval: 60000,
		// daily
		dataUpdateInterval: 86400000,
		// show next 5 transports
		display: 10,

		use24HrClock: true,

		stops: [],
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	parsedData: [],

	dayColumns: ["sunday","monday","tuesday","wednesday","thursday","friday","saturday"],

	routeTypesMap: {
		0: "<i class=\"fas fa-train\"></i>",	// tram, light-rail
		1: "<i class=\"fas fa-subway\"></i>",	// subway
		2: "<i class=\"fas fa-train\"></i>",	// long distance rail
		3: "<i class=\"fas fa-bus-alt\"></i>",	// bus
		4: "<i class=\"fas fa-ship\"></i>",		// ferry
		5: "<i class=\"fas fa-tram\"></i>",		// cable tram
		6: "<i class=\"fas fa-tram\"></i>",		// aerial lift
		7: "<i class=\"fas fa-train\"></i>",	// funicular
		11: "<i class=\"fas fa-bus-alt\"></i>",	// trolleybus
		12: "<i class=\"fas fa-train\"></i>",	// monorail
		800: "<i class=\"fas fa-bus-alt\"></i>",// trolleybus, blame "Rīgas satiksme" for this
		900: "<i class=\"fas fa-train\"></i>",	// tram, blame "Rīgas satiksme" for this
	},

	dataNotification: [],

	getScripts: function() {
		return [];
	},

	getStyles: function () {
		return [
			"MMM-public-transport-gtfs.css",
			"font-awesome.css",
		];
	},

	// Load translations files
	getTranslations: function() {
		return {
			en: "translations/en.json",
		};
	},

	// socketNotificationReceived from helper
	socketNotificationReceived: function (notification, payload) {
		if (notification === this.name + "-SET_GTFS") {
			this.processData(payload);
		} else if (notification === this.name + "-SET_NOTIFICATIONS") {
			this.dataNotification.push(payload);
		}
	},

	start: function() {
		var self = this;
		//Flag for check if module is loaded
		this.loaded = false;

		// Schedule update timer.
		this.getData();

		setInterval(function() {
			self.updateDom();
		}, this.config.domUpdateInterval);
	},

	getData: function() {
		this.sendSocketNotification(this.name + "-GET_GTFS", {stops: this.config.stops, url: this.config.gtfsDataUrl});
	},

	scheduleUpdate: function() {
		var nextLoad = this.config.dataUpdateInterval;
		var self = this;
		setTimeout(function() {
			self.getData();
		}, nextLoad);
	},

	getDom: function() {
		var self = this;
		// create element wrapper for show into the module
		var wrapper = document.createElement("div");
		// If this.dataRequest is not empty
		if (this.dataRequest) {
			var wrapperDataRequest = document.createElement("div");

			var content = "";
			var currentTime = new Date();

			var nextStops = [];

			var chooseNextStops = function(stopTime, index, array){
				if (
					stopTime.departure_time.getMinutes() > currentTime.getMinutes()
					|| stopTime.departure_time.getHours() > currentTime.getHours()
				) {
					var trip = self.dataRequest.trips[stopTime.trip_id];
					var calendar = self.dataRequest.serviceCalendar[trip.service_id];
					var calendarExceptions = self.dataRequest.calendarExceptions[trip.service_id];
					var hasCalendar = typeof calendar !== "undefined";

					if (
						(!hasCalendar || self.isInTimePeriod(currentTime, calendar.start_date, calendar.end_date))
						&& (
							(hasCalendar && calendar[self.dayColumns[currentTime.getDay()]] === 1)
							|| self.isEnabledByException(currentTime, calendarExceptions)
						)
						&& !self.isDisabledByExceptions(currentTime, calendarExceptions)
					) {
						nextStops.push(stopTime);
					}
				}

				return nextStops.length < self.config.display;
			};

			for(var i = 0; i < 24; i++) {
				var checkingHour = currentTime.getHours() + (i > 0 ? 1 : 0);
				currentTime.setMinutes((i > 0 ? 0 : currentTime.getMinutes()));
				if (checkingHour > 23){
					currentTime.setDate(currentTime.getDate() + 1);
					checkingHour = 0;
				}

				currentTime.setHours(checkingHour);

				if (typeof this.dataRequest.stopTimes[checkingHour] !== "undefined") {
					this.dataRequest.stopTimes[checkingHour].every(chooseNextStops);
				}

				if (nextStops.length >= this.config.display) {
					break;
				}
			}

			nextStops.forEach(function(nextStop, index, array){
				var trip = self.dataRequest.trips[nextStop.trip_id]
				var route = self.dataRequest.routes[trip.route_id];
				content += "<table class='small align-left'>";
				content += "<tr>";
				content += 		"<th class='detail detail-route' colspan='2'>";
				content +=			"<span class='detail-span'>";
				content += 				(typeof route.route_color !== "undefined" ? "<span style='color: #" + route.route_color + "'>" : "")
											+ self.routeTypesMap[route.route_type]
										+ (typeof route.route_color !== "undefined" ? "</span>" : "")
										+ " " + route.route_short_name + " : " + route.route_long_name;
				content +=			"</span>";
				content += 		"</th>";
				content += "</tr>";
				content += "<tr>";
				content += 		"<td class='detail detail-departure-time align-center " + (!self.config.use24HrClock ? "hr12-clock" : "") + "' rowspan='3'>";
				content +=			"<span class='detail-span'>";
				content += 				self.getClockTimeString(nextStop.departure_time);
				content +=			"</span>";
				content += 		"</td>";
				content += "</tr>";
				content += "<tr>";
				content += 		"<td class='detail detail-stop'>";
				content +=			"<span class='detail-span'>";
				content += 				"<i class=\"fas fa-map-marker-alt\"></i> " + self.dataRequest.stops[nextStop.stop_id].stop_name;
				content +=			"</span>";
				content += 		"</td>";
				content += "</tr>";
				content += "<tr>";
				content += 		"<td class='detail detail-direction'>";
				content +=			"<span class='detail-span'>";
				content += 				"<i class=\"fas fa-map-pin\"></i> " + trip.trip_headsign
				content +=			"</span>";
				content += 		"</td>";
				content += "</tr>";
				content += "</table>";
			});

			wrapperDataRequest.innerHTML = content;

			wrapper.appendChild(wrapperDataRequest);
		}

		// Data from helper
		if (this.dataNotification.length > 0) {
			var wrapperDataNotification = document.createElement("div");
			this.dataNotification.forEach(function(notification, index, array) {
				wrapperDataNotification.innerHTML +=  self.translate(notification.code) + " " + notification.additionalData.replace(/\\n/g, "\n") + "<br>";
			});
			wrapper.appendChild(wrapperDataNotification);
		}
		return wrapper;
	},

	processData: function(data) {
		// in notification Date gets converted to string... converting it back to date for convenience
		Object.keys(data.stopTimes).forEach(function(hour, index, array) {
			data.stopTimes[hour].forEach(function(stopTime, index, array) {
				data.stopTimes[hour][index].departure_time = new Date(stopTime.departure_time);
				data.stopTimes[hour][index].arrival_time = new Date(stopTime.arrival_time);
			});
		});

		Object.keys(data.serviceCalendar).forEach(function(serviceId, index, array) {
			data.serviceCalendar[serviceId].start_date = new Date(data.serviceCalendar[serviceId].start_date);
			data.serviceCalendar[serviceId].end_date = new Date(data.serviceCalendar[serviceId].end_date);
		});

		Object.keys(data.calendarExceptions).forEach(function(serviceId, index, array) {
			data.calendarExceptions[serviceId].forEach(function(calendarException, index, array) {
				data.calendarExceptions[serviceId][index].date = new Date(data.calendarExceptions[serviceId][index].date);
			});
		});

		this.dataRequest = data;
		Log.info(this.name, "GTFS data processed");
		Log.debug(this.name, "GTFS data", data);

		if (this.loaded === false) { this.updateDom(this.config.animationSpeed); }
		this.loaded = true;
		this.scheduleUpdate();
	},

	isInTimePeriod(currentTime, startTime, endTime) {
		return startTime <= currentTime
			&& endTime >= currentTime;
	},

	isEnabledByException(currentTime, calendarExceptions) {
		// if no exceptions - great!
		if (typeof calendarExceptions === "undefined" || calendarExceptions === null) {
			return false;
		}

		var exceptionOutcome = false;

		calendarExceptions.every(function(calendarException, index, array) {
			// if exception is today
			if(
				calendarException.date.getDate() === currentTime.getDay()
				&& calendarException.date.getMonth() === currentTime.getMonth()
				&& calendarException.date.getFullYear() === currentTime.getFullYear()
			) {
				exceptionOutcome = calendarException.exception_type === "1"; // 1 - new service just for this date, 2 - cancelled for today
				return false;
			}
		});

		return exceptionOutcome;
	},

	isDisabledByExceptions(currentTime, calendarExceptions) {
		// if no exceptions - great!
		if (typeof calendarExceptions === "undefined" || calendarExceptions === null) {
			return false;
		}

		var exceptionOutcome = false;

		calendarExceptions.every(function(calendarException, index, array) {
			// if exception is today
			if(
				calendarException.date.getDate() === currentTime.getDay()
				&& calendarException.date.getMonth() === currentTime.getMonth()
				&& calendarException.date.getFullYear() === currentTime.getFullYear()
			) {
				exceptionOutcome = calendarException.exception_type === "2"; // 1 - new service just for this date, 2 - cancelled for today
				return false;
			}
		});

		return exceptionOutcome;
	},

	getDoubleDigitNumber(number) {
		while ((number + "").length < 2) {
			number = "0" + number;
		}

		return number;
	},

	// Yes, I could not find format method for date in JS *confused screeching*
	getClockTimeString(date) {
		var hours = date.getHours();
		var pm = "";
		if (!this.config.use24HrClock) {
			if (hours > 12) {
				hours -= 12;
			}

			if (hours < 11) {
				pm = " PM";
			}else {
				pm = " AM";
			}
		}

		return this.getDoubleDigitNumber(hours) + ":" + this.getDoubleDigitNumber(date.getMinutes()) + pm;
	}
});


MMM-public-transport-schedule.js
Module.register("MMM-public-transport-gtfs", {
	defaults: {
		gtfsDataUrl: "",
		// every minute
		domUpdateInterval: 60000,
		// daily
		dataUpdateInterval: 86400000,
		// show next 5 transports
		display: 10,

		use24HrClock: true,

		stops: [],
	},

	requiresVersion: "2.1.0", // Required version of MagicMirror

	parsedData: [],

	dayColumns: ["sunday","monday","tuesday","wednesday","thursday","friday","saturday"],

	routeTypesMap: {
		0: "<i class=\"fas fa-train\"></i>",	// tram, light-rail
		1: "<i class=\"fas fa-subway\"></i>",	// subway
		2: "<i class=\"fas fa-train\"></i>",	// long distance rail
		3: "<i class=\"fas fa-bus-alt\"></i>",	// bus
		4: "<i class=\"fas fa-ship\"></i>",		// ferry
		5: "<i class=\"fas fa-tram\"></i>",		// cable tram
		6: "<i class=\"fas fa-tram\"></i>",		// aerial lift
		7: "<i class=\"fas fa-train\"></i>",	// funicular
		11: "<i class=\"fas fa-bus-alt\"></i>",	// trolleybus
		12: "<i class=\"fas fa-train\"></i>",	// monorail
		800: "<i class=\"fas fa-bus-alt\"></i>",// trolleybus, blame "Rīgas satiksme" for this
		900: "<i class=\"fas fa-train\"></i>",	// tram, blame "Rīgas satiksme" for this
	},

	dataNotification: [],

	getScripts: function() {
		return [];
	},

	getStyles: function () {
		return [
			"MMM-public-transport-gtfs.css",
			"font-awesome.css",
		];
	},

	// Load translations files
	getTranslations: function() {
		return {
			en: "translations/en.json",
		};
	},

	// socketNotificationReceived from helper
	socketNotificationReceived: function (notification, payload) {
		if (notification === this.name + "-SET_GTFS") {
			this.processData(payload);
		} else if (notification === this.name + "-SET_NOTIFICATIONS") {
			this.dataNotification.push(payload);
		}
	},

	start: function() {
		var self = this;
		//Flag for check if module is loaded
		this.loaded = false;

		// Schedule update timer.
		this.getData();

		setInterval(function() {
			self.updateDom();
		}, this.config.domUpdateInterval);
	},

	getData: function() {
		this.sendSocketNotification(this.name + "-GET_GTFS", {stops: this.config.stops, url: this.config.gtfsDataUrl});
	},

	scheduleUpdate: function() {
		var nextLoad = this.config.dataUpdateInterval;
		var self = this;
		setTimeout(function() {
			self.getData();
		}, nextLoad);
	},

	getDom: function() {
		var self = this;
		// create element wrapper for show into the module
		var wrapper = document.createElement("div");
		// If this.dataRequest is not empty
		if (this.dataRequest) {
			var wrapperDataRequest = document.createElement("div");

			var content = "";
			var currentTime = new Date();

			var nextStops = [];

			var chooseNextStops = function(stopTime, index, array){
				if (
					stopTime.departure_time.getMinutes() > currentTime.getMinutes()
					|| stopTime.departure_time.getHours() > currentTime.getHours()
				) {
					var trip = self.dataRequest.trips[stopTime.trip_id];
					var calendar = self.dataRequest.serviceCalendar[trip.service_id];
					var calendarExceptions = self.dataRequest.calendarExceptions[trip.service_id];
					var hasCalendar = typeof calendar !== "undefined";

					if (
						(!hasCalendar || self.isInTimePeriod(currentTime, calendar.start_date, calendar.end_date))
						&& (
							(hasCalendar && calendar[self.dayColumns[currentTime.getDay()]] === 1)
							|| self.isEnabledByException(currentTime, calendarExceptions)
						)
						&& self.isNotDisabledByExceptions(currentTime, calendarExceptions)
					) {
						nextStops.push(stopTime);
					}
				}

				return nextStops.length < self.config.display;
			};

			for(var i = 0; i < 24; i++) {
				var checkingHour = currentTime.getHours() + (i > 0 ? 1 : 0);
				currentTime.setMinutes((i > 0 ? 0 : currentTime.getMinutes()));
				if (checkingHour > 23){
					currentTime.setDate(currentTime.getDate() + 1);
					checkingHour = 0;
				}

				currentTime.setHours(checkingHour);

				if (typeof this.dataRequest.stopTimes[checkingHour] !== "undefined") {
					this.dataRequest.stopTimes[checkingHour].every(chooseNextStops);
				}

				if (nextStops.length >= this.config.display) {
					break;
				}
			}

			nextStops.forEach(function(nextStop, index, array){
				var trip = self.dataRequest.trips[nextStop.trip_id]
				var route = self.dataRequest.routes[trip.route_id];
				content += "<table class='small align-left'>";
				content += "<tr>";
				content += 		"<th class='detail detail-route' colspan='2'>";
				content +=			"<span class='detail-span'>";
				content += 				(typeof route.route_color !== "undefined" ? "<span style='color: #" + route.route_color + "'>" : "")
											+ self.routeTypesMap[route.route_type]
										+ (typeof route.route_color !== "undefined" ? "</span>" : "")
										+ " " + route.route_short_name + " : " + route.route_long_name;
				content +=			"</span>";
				content += 		"</th>";
				content += "</tr>";
				content += "<tr>";
				content += 		"<td class='detail detail-departure-time align-center " + (!self.config.use24HrClock ? "hr12-clock" : "") + "' rowspan='3'>";
				content +=			"<span class='detail-span'>";
				content += 				self.getClockTimeString(nextStop.departure_time);
				content +=			"</span>";
				content += 		"</td>";
				content += "</tr>";
				content += "<tr>";
				content += 		"<td class='detail detail-stop'>";
				content +=			"<span class='detail-span'>";
				content += 				"<i class=\"fas fa-map-marker-alt\"></i> " + self.dataRequest.stops[nextStop.stop_id].stop_name;
				content +=			"</span>";
				content += 		"</td>";
				content += "</tr>";
				content += "<tr>";
				content += 		"<td class='detail detail-direction'>";
				content +=			"<span class='detail-span'>";
				content += 				"<i class=\"fas fa-map-pin\"></i> " + trip.trip_headsign
				content +=			"</span>";
				content += 		"</td>";
				content += "</tr>";
				content += "</table>";
			});

			wrapperDataRequest.innerHTML = content;

			wrapper.appendChild(wrapperDataRequest);
		}

		// Data from helper
		if (this.dataNotification.length > 0) {
			var wrapperDataNotification = document.createElement("div");
			this.dataNotification.forEach(function(notification, index, array) {
				wrapperDataNotification.innerHTML +=  self.translate(notification.code) + " " + notification.additionalData.replace(/\\n/g, "\n") + "<br>";
			});
			wrapper.appendChild(wrapperDataNotification);
		}
		return wrapper;
	},

	processData: function(data) {
		// in notification Date gets converted to string... converting it back to date for convenience
		Object.keys(data.stopTimes).forEach(function(hour, index, array) {
			data.stopTimes[hour].forEach(function(stopTime, index, array) {
				data.stopTimes[hour][index].departure_time = new Date(stopTime.departure_time);
				data.stopTimes[hour][index].arrival_time = new Date(stopTime.arrival_time);
			});
		});

		Object.keys(data.serviceCalendar).forEach(function(serviceId, index, array) {
			data.serviceCalendar[serviceId].start_date = new Date(data.serviceCalendar[serviceId].start_date);
			data.serviceCalendar[serviceId].end_date = new Date(data.serviceCalendar[serviceId].end_date);
		});

		Object.keys(data.calendarExceptions).forEach(function(serviceId, index, array) {
			data.calendarExceptions[serviceId].forEach(function(calendarException, index, array) {
				data.calendarExceptions[serviceId][index].date = new Date(data.calendarExceptions[serviceId][index].date);
			});
		});

		this.dataRequest = data;
		Log.info(this.name, "GTFS data processed");
		Log.debug(this.name, "GTFS data", data);

		if (this.loaded === false) { this.updateDom(this.config.animationSpeed); }
		this.loaded = true;
		this.scheduleUpdate();
	},

	isInTimePeriod(currentTime, startTime, endTime) {
		return startTime <= currentTime
			&& endTime >= currentTime;
	},

	isEnabledByException(currentTime, calendarExceptions) {
		// if no exceptions - great!
		if (typeof calendarExceptions === "undefined" || calendarExceptions === null) {
			return false;
		}

		var exceptionOutcome = false;

		calendarExceptions.every(function(calendarException, index, array) {
			// if exception is today
			if(
				calendarException.date.getDate() === currentTime.getDay()
				&& calendarException.date.getMonth() === currentTime.getMonth()
				&& calendarException.date.getFullYear() === currentTime.getFullYear()
			) {
				exceptionOutcome = calendarException.exception_type === "1"; // 1 - new service just for this date, 2 - cancelled for today
				return false;
			}
		});

		return exceptionOutcome;
	},

	isNotDisabledByExceptions(currentTime, calendarExceptions) {
		// if no exceptions - great!
		if (typeof calendarExceptions === "undefined" || calendarExceptions === null) {
			return true;
		}

		var exceptionOutcome = true;

		calendarExceptions.every(function(calendarException, index, array) {
			// if exception is today
			if(
				calendarException.date.getDate() === currentTime.getDay()
				&& calendarException.date.getMonth() === currentTime.getMonth()
				&& calendarException.date.getFullYear() === currentTime.getFullYear()
			) {
				exceptionOutcome = calendarException.exception_type !== "1"; // 1 - new service just for this date, 2 - cancelled for today
				return false;
			}
		});

		return exceptionOutcome;
	},

	getDoubleDigitNumber(number) {
		while ((number + "").length < 2) {
			number = "0" + number;
		}

		return number;
	},

	// Yes, I could not find format method for date in JS *confused screeching*
	getClockTimeString(date) {
		var hours = date.getHours();
		var pm = "";
		if (!this.config.use24HrClock) {
			if (hours > 12) {
				hours -= 12;
			}

			if (hours < 11) {
				pm = " PM";
			}else {
				pm = " AM";
			}
		}

		return this.getDoubleDigitNumber(hours) + ":" + this.getDoubleDigitNumber(date.getMinutes()) + pm;
	}
});
